Add category system to plugin configuration UI. Fixes #2411

AdamBallinger opened the pull request:
![001447b626d2d6324d0fbc55c2c883f5](https://user-images.githubusercontent.com/5104566/39702442-ba9d674e-51fc-11e8-9225-7f41d70705f4.gif)
Adds category field to PluginDescriptor annotation.
New categories can be added in the PluginCategory enum.
New categories are automatically added to the config UI.

deathbeam commented:
Well, I must say, this looks horrible. Maybe a dropdown instead?

deathbeam commented:
Also, both utility and other should not be categories, that is basically uncategorized stuff that you dumped somewhere, so if you want to properly categorize them, remove these 2 and figure out something better.

AdamBallinger commented:
Thanks for the feedback! Will get started on improvements.

AdamBallinger commented:
Removed other category and replaced it with client and updated plugins accordingly. I think utility as a category is fine as plugins that use the utility category are exactly that; utility.
![001447b626d2d6324d0fbc55c2c883f5](https://user-images.githubusercontent.com/5104566/39707141-6e9568c4-520b-11e8-829d-a20f4d9d3bd0.gif)

deathbeam commented:
It still feels like having everything that do not fit anywhere in utility and client. I would add Items categories, then Minigames, Bosses, Indicators, Combat looks fine I guess.

AdamBallinger commented:
Added 3 new categories and I feel like what remains in Utility / Client are hard to really categorise as anything else.

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/ClueScrollPlugin.java:
```diff
@@ -87,7 +87,7 @@
 @PluginDescriptor(
 	name = "Clue Scroll",
-	category = PluginCategory.UTILITY
+	category = PluginCategory.ACTIVITY
```
Clue scrolls are minigame

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/implings/ImplingsPlugin.java:
```diff
@@ -49,7 +49,7 @@
 @PluginDescriptor(
 	name = "Implings",
-	category = PluginCategory.UTILITY
+	category = PluginCategory.ACTIVITY
```
Implings are minigame

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/kourendlibrary/KourendLibraryPlugin.java:
```diff
@@ -52,7 +52,7 @@
 @PluginDescriptor(
 	name = "Kourend Library",
-	category = PluginCategory.UTILITY
+	category = PluginCategory.ACTIVITY
```
Isn't this minigame too? Sort of

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java:
```diff
@@ -149,11 +151,21 @@ final void rebuildPluginList()
 				.sorted(Comparator.comparing(left -> left.getClass().getAnnotation(PluginDescriptor.class).name()))
 				.forEach(plugin ->
 				{
+					final PluginCategory pluginCategory = plugin.getClass()
```
Get the annotation only once and use it for both category and plugin name

AdamBallinger commented:
![531fb1ac7132823a29941a7a6e45d910](https://user-images.githubusercontent.com/5104566/39961323-583d9028-562b-11e8-88a6-765af08f7bb7.gif)
I have removed the dropdown box and instead display a label for each category with a list of plugins under it that are assigned to that category.
I have also removed the ALL category and with it the default category for PluginDescriptor.

AdamBallinger commented:
As the obsidian theme was merged I have made changes to make the style of the category labels consistent with the rest of the theme. 
![ae80ac83c5d7984e85d35676d8f5e73e](https://user-images.githubusercontent.com/5104566/40272912-1ef35af4-5bae-11e8-8506-413152ee88fc.png)
Let me know if it can be improved in any way.
